app/decorators/user_decorator.rb

```diff
@@ -61,6 +61,46 @@ def no_longer_blocked_from_entering_2fa_code?
     user.second_factor_locked_at && blocked_from_2fa_period_expired?
   end
 
+  def slo_identity(saml_request)
+    return first_identity if
+            saml_request &&
+            saml_request.issuer != first_identity.service_provider
+    # Logout was initiated out of chronological order. Resume SLO
+    # with next Identity
+    return second_identity if second_identity
+    # fallback to last authenticated identity (for 3+ SP envs)
+    last_identity
+  end
+
+  def first_identity
+    active_identities[0]
+  end
+
+  def second_identity
+    active_identities[1]
+  end
+
+  def last_identity
```

Comment:
This stuff kinda feels like it *is* part of the data model to me, since its about the user->identity data relation, but maybe I'm wrong?  Convince me :)